segmentstore/server/src/main/java/io/pravega/segmentstore/server/attributes/SegmentAttributeIndex.java

```diff
@@ -0,0 +1,663 @@
+                            b.put(attributeId, value);
+                        }
+                    });
+                    return b.build();
+                });
+    }
+
+    @Override
+    public CompletableFuture<Long> get(UUID key, Duration timeout) {
+        ensureInitialized();
+        return readAllSinceLastSnapshot(timeout)
+                .thenApply(c -> c.attributes.get(key));
+    }
+
+    @Override
+    public CompletableFuture<Void> remove(UUID key, Duration timeout) {
+        return remove(Collections.singleton(key), timeout);
+    }
+
+    @Override
+    public CompletableFuture<Void> remove(Collection<UUID> keys, Duration timeout) {
+        Preconditions.checkNotNull(keys, "keys");
+        return put(keys.stream().collect(Collectors.toMap(key -> key, key -> Attributes.NULL_ATTRIBUTE_VALUE)), timeout);
+    }
+
+    @Override
+    public CompletableFuture<Void> seal(Duration timeout) {
+        ensureInitialized();
+        TimeoutTimer timer = new TimeoutTimer(timeout);
+        return Futures.exceptionallyExpecting(
+                createSnapshot(new AttributeCollection(), true, timer.getRemaining())
+                        .thenComposeAsync(v -> this.storage.seal(this.attributeSegment.get().handle, timer.getRemaining()), this.executor)
+                        .thenRun(() -> log.info("{}: Sealed (Length = {}).", this.traceObjectId, this.attributeSegment.get().getLength())),
+                ex -> ex instanceof StreamSegmentSealedException,
+                null);
+    }
+
+    //endregion
+
+    //region Operations
+
+    /**
+     * Determines if we should generate a Snapshot based on the current state of the Attribute Segment.
+     *
+     * @return True if we should generate a Snapshot, false otherwise.
+     */
+    @VisibleForTesting
+    boolean shouldSnapshot() {
+        long lastSnapshotEndOffset = getLastSnapshotOffset() + getLastSnapshotLength();
+        return this.attributeSegment.get().getLength() - lastSnapshotEndOffset >= this.config.getSnapshotTriggerSize();
+    }
+
+    /**
+     * Reads all attributes beginning with the last recorded snapshot until the current end of the Attribute Segment.
+     *
+     * @param timeout Timeout for the operation.
+     * @return A CompletableFuture that, when completed, will contain the Attributes as they should be when reaching the
+     * end of the Attribute Segment.
+     */
+    private CompletableFuture<AttributeCollection> readAllSinceLastSnapshot(Duration timeout) {
+        CompletableFuture<AttributeCollection> result = READ_RETRY.runAsync(() -> {
+            if (isMainSegmentDeleted()) {
+                // Verify if the main segment has been deleted and stop now if needed.
+                return handleMainSegmentDeleted(timeout);
+            }
+
+            long lastSnapshotOffset = getLastSnapshotOffset();
+            int readLength = (int) Math.min(Integer.MAX_VALUE, this.attributeSegment.get().getLength() - lastSnapshotOffset);
```

Comment:
Does this yields that a single snapshot can be at most 2.14GB in size? If this is correct, could it be interesting to mention this limitation at some point of the config/doc?

Reply:
Each attribute-value pair in the snapshot takes 24 bytes. IF we ever reach the maximum then we'll have about 100M attributes. We'll run into trouble way before that reaches. This whole thing is designed to work well with up to 1M attributes (for which the snapshot will be about 24MB).